pkg/provider/onepassword/onepassword.go

```diff
@@ -0,0 +1,364 @@
+	}
+
+	err := kube.Get(ctx, objectKey, credentialsSecret)
+	if err != nil {
+		return nil, fmt.Errorf(errFetchK8sSecret, err)
+	}
+	token := credentialsSecret.Data[config.Auth.SecretRef.ConnectToken.Key]
+	if (token == nil) || (len(token) == 0) {
+		return nil, fmt.Errorf(errMissingToken)
+	}
+	provider.client = connect.NewClientWithUserAgent(config.ConnectHost, string(token), userAgent)
+	provider.vaults = config.Vaults
+
+	return provider, nil
+}
+
+// ValidateStore checks if the provided store is valid.
+func (provider *ProviderOnePassword) ValidateStore(store esv1beta1.GenericStore) error {
+	return validateStore(store)
+}
+
+func validateStore(store esv1beta1.GenericStore) error {
+	storeSpec := store.GetSpec()
+	if storeSpec == nil || storeSpec.Provider == nil || storeSpec.Provider.OnePassword == nil {
+		return fmt.Errorf(errOnePasswordStore)
+	}
+	config := storeSpec.Provider.OnePassword
+
+	// only ClusterSecretStore is allowed to set namespace (and then it's required)
+	if store.GetObjectKind().GroupVersionKind().Kind == esv1beta1.ClusterSecretStoreKind {
+		if config.Auth.SecretRef.ConnectToken.Namespace == nil {
+			return fmt.Errorf(errClusterSecretStoreNamespaceMissing)
+		}
+	} else {
+		if config.Auth.SecretRef.ConnectToken.Namespace != nil {
+			return fmt.Errorf(errSecretStoreNamespaceSet)
+		}
+	}
+
+	return nil
+}
+
+// GetSecret returns a single secret from the provider.
+func (provider *ProviderOnePassword) GetSecret(ctx context.Context, ref esv1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	item, err := provider.findItem(ref.Key)
+	if err != nil {
+		return nil, err
+	}
+
+	// handle files
+	if item.Category == documentCategory {
+		for _, file := range item.Files {
+			// default to the first file when ref.Property is empty
+			if file.Name == ref.Property || ref.Property == "" {
+				contents, err := provider.client.GetFileContent(file)
+				if err != nil {
+					return nil, err
+				}
+
+				return contents, nil
+			}
+		}
+
+		return nil, fmt.Errorf(errDocumentNotFound, fmt.Errorf("'%s', '%s'", item.Title, ref.Property))
+	}
+
+	// handle fields
+	// default to a field labeled `password`
+	fieldLabel := "password"
+	if ref.Property != "" {
+		fieldLabel = ref.Property
+	}
+
+	if length := countFieldsWithLabel(fieldLabel, item.Fields); length != 1 {
+		return nil, fmt.Errorf(errExpectedOneField, fmt.Errorf("'%s' in '%s', got %d", fieldLabel, item.Title, length))
+	}
+
+	// caution: do not use client.GetValue here because it has undesirable behavior on keys with a dot in them
+	value := ""
+	for _, field := range item.Fields {
+		if field.Label == fieldLabel {
+			value = field.Value
+			break
+		}
+	}
+
+	return []byte(value), nil
+}
+
+// Validate checks if the client is configured correctly
+// to be able to retrieve secrets from the provider.
+func (provider *ProviderOnePassword) Validate() error {
+	for vaultName := range provider.vaults {
```

Comment:
I would move this check to validateStore(). And check here something like actually using the client to list vaults or something like that (or using an appropriate method for checking the connection, if the SDK provides that).

Reply:
`provider.client.GetItems(vaultName)` is actually using the 1Password client to confirm client access. I set it it up to check access for each vault configured. The API doesn't have a `ping` endpoint. I chose `GetItems` because it seemed like the least complex call to make in lieu of a ping endpoint.

Let me know if I misunderstood or anything.

Reply:
My bad, I missed the call to getitems when looking before. It makes sense :)